apple/internal/resource_rules/apple_metal_library.bzl

```diff
@@ -0,0 +1,182 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of Metal shader library rule."""
+
+load(
+    "@bazel_skylib//lib:dicts.bzl",
+    "dicts",
+)
+load(
+    "@bazel_skylib//lib:paths.bzl",
+    "paths",
+)
+load(
+    "@build_bazel_apple_support//lib:apple_support.bzl",
+    "apple_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:apple_toolchains.bzl",
+    "AppleXPlatToolsToolchainInfo",
+    "apple_toolchain_utils",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:features_support.bzl",
+    "features_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:platform_support.bzl",
+    "platform_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:resources.bzl",
+    "resources",
+)
+
+def _metal_apple_target_triple(platform_prerequisites):
+    """Returns a `metal` compiler-compatible target triple string.
+
+    Args:
+      platform_prerequisites: Struct containing information on the platform being targeted.
+
+    Returns:
+      A Metal target triple string for use with the `-target` copt.
+    """
+    target_os_version = platform_prerequisites.minimum_os
+
+    platform = platform_prerequisites.apple_fragment.single_arch_platform
+    platform_string = str(platform.platform_type)
+    if platform_string == "macos":
+        platform_string = "macosx"
+
+    environment = "" if platform.is_device else "-simulator"
+
+    return "air64-apple-{platform}{version}{environment}".format(
```

Comment:
air64?

Reply:
yup! metal uses an underdocumented specialized target triple format for individual compiled shaders (`.air`) that look like LLVM triples but are decidedly not. this approach is used to harmonize better with platform transitions in rules_apple, assuming i'm understanding things correctly.